### ISSUE TITLE
Beginners documentation

### ISSUE THREAD

**Issue by hanckmann** [redox-os/orbtk#106]
# Context

There is very (too?) little documentation available, especially for newbies.

# Problem description & Solution

Documentation can be generated with: `cargo doc --no-deps --open`
The resulting documentation is nice and all, but a bit to daunting for a beginner.
The examples are super helpful, but do not provide sufficient context.
It would be helpful to have some documentation on how to get started. How to connect functions to buttons, update labels (textboxes), etc. These can be extracted from the examples, but only with a lot of effort.

# Help?
I am going to try to get some stuff made with orbTk. I would really love to help out by writing documentation (a starters guide or something... to help beginners). However, I would be helpful if somebody could create a start on which I can continue...

**Comment by FloVanGH** [comment]
@hanckmann git push --set-upstream origin 147-book. I've started with an initial version of the OrbTk book (check #107 and #108). If you like you could write the chapter "Write your first app" if you like. I could also assign you to the related issue and pull request.

**Comment by hanckmann** [comment]
FYI.
I have started with the "first app" chapter.
I must admit that this is the first time doing this. So you might have to help me out here.
I have forked the repo, checked out the 147-book branch, and editing and committing my changes there.
I will ask later how to get those changes into your repository via a pull request (which I have also never done before).

**Comment by FloVanGH** [comment]
That’s great. Did you find the files of the book? A info bar at the top your repository should shown if you want to create a pull request for sour branch.

**Comment by SerhiiBilyk** [comment]
@hanckmann @FloVanGH Hi guys, I would like to help with book. I made a little research and I found that the orbrk is the most understandable GUI framework for me and it just work. What I can't say about other frameworks.

**Comment by FloVanGH** [comment]
@SerhiiBilyk @hanckmann is working on the chapter write your first app. If you are interested you could start reading the book and give us feedback. If you miss something you could add it if you like. Maybe you could start with a second chapter. What do you think @hanckmann?

**Comment by SerhiiBilyk** [comment]
@hanckmann 
```rust
use orbtk::prelude::*;

#[derive(Debug, Copy, Clone)]
enum Action {
    Increment(i32),
}

#[derive(Default)]
pub struct MainViewState {
    num: i32,
    action: Option<Action>,
}

impl MainViewState {
    fn action(&mut self, action: impl Into<Option<Action>>) {
        self.action = action.into();
    }
}

impl Template for MainView {
    fn template(self, id: Entity, ctx: &mut BuildContext) -> Self {
        self.name("MainView").child(
            Button::create()
                .min_size(100.0, 50.0)
                .text(String::from("hello").to_string())
                .on_click(move |states| -> bool {
                    println!("hello {:?}", states);
                    //  state(id, states).action(Action::Increment(10));

                    //  state(id, states).action(Action::Operator(sight));
                    true
                })
                .build(ctx),
        )
    }
}

impl State for MainViewState {
    fn update(&self, ctx: &mut Context) {
        if let Some(action) = self.action {
            match action {
                Action::Increment(digit) => {
                    //    self.num += digit;
                    println!("Self {:?}", self.num);
                }

                _ => {}
            }
        }

        //  self.action = None;
    }
}

widget!(MainView<MainViewState> {});

fn main() {
    Application::new()
        .window(|ctx| {
            Window::create()
                .title("OrbTk - minimal example")
                .position((100.0, 100.0))
                .size(400.0, 400.0)
                .child(MainView::create().build(ctx))
                .build(ctx)
        })
        .run();
}
/// StatesContext
fn state<'a>(id: Entity, states: &'a mut StatesContext) -> &'a mut MainViewState {
    states.get_mut(id)
}
```
How can I import StatesContext ? 
I cant find it even in documentation.

**Comment by FloVanGH** [comment]
@hanckmann which dependency do you use in your Cargo.toml file `orbtk="0.3.1-alpha1"` or `orbtk= { git = "https://github.com/redox-os/orbtk.git", branch="develop" }`. You have to use the git version, StatesContext is a new feature and not now available on the crates.io version.

**Comment by SerhiiBilyk** [comment]
I made a simple  [counter](https://github.com/SerhiiBilyk/gui/blob/master/src/main.rs)
But, I don't know/understand how to bind MainView with state. See [here](https://github.com/SerhiiBilyk/gui/blob/master/src/main.rs#L38)
Can we get rid of [this](https://github.com/SerhiiBilyk/gui/blob/master/src/main.rs#L63) ? There is  an infinity loop without this line

**Comment by SerhiiBilyk** [comment]
@FloVanGH is [this](https://github.com/SerhiiBilyk/gui/blob/master/src/main.rsl) ok for minimum counter example?

**Comment by FloVanGH** [comment]
@SerhiiBilyk the link is broken.

**Comment by SerhiiBilyk** [comment]
@FloVanGH https://github.com/SerhiiBilyk/gui/blob/master/src/main.rs

**Comment by FloVanGH** [comment]
@SerhiiBilyk looks great. 

Here some feedback:

```rust
// Is it possible to get rid of this line ?
self.action = None;
```
Yes it is. Check the work in progress todo app example: https://github.com/redox-os/orbtk/pull/109/files#diff-88def14d15394829e2ec6d406d7342a7R13. The App state uses a VecDeque for the actions / events. So you can pop all incoming actions from the vec deque instead of have an single action and set `self.action = None;`

The window title should be changed to `Counter` or `Write your first app` or what you think is the right name for your example. 

I think you could remove the num field from the state.

Thank you.

**Comment by SerhiiBilyk** [comment]
@FloVanGH Thank you

**Comment by clouds56** [comment]
I'm beginner and wonder when the `update` would be called? Why `self.action = Some(action)` would work if 2 actions comes simultaneously?

**Comment by FloVanGH** [comment]
@clouds56 in one run the update method of a state is called one time. If the state should handle multiple actions depends on the use case. If the state should handle only handle one action at time a single action is enough, otherwise e.g. a `VecDeq<Action>` could be used.

**Comment by clouds56** [comment]
So if I understand correctly, when a any events (like click, keyboard) arrives, the corresponding widget would notified by handler like `on_click`, after all widgets responses, it would call `update` on **all** widgets (but not only the widgets affected by the event). And this is "one run" per event, right?
Therefore in the calculator example, 2 clicks must trigger 2 runs, so that there's no action would be lost.

I'd like to read the corresponding source code if possible but I'm not managed to find the location.

**Comment by FloVanGH** [comment]
@clouds56 yes you understand right. The source code you are looking for is in the EventStateSystem https://github.com/redox-os/orbtk/blob/develop/crates/api/src/systems/event_state_system.rs.

**Comment by rzerres** [comment]
> FYI.
> I have started with the "first app" chapter.
> I must admit that this is the first time doing this. So you might have to help me out here.
> I have forked the repo, checked out the 147-book branch, and editing and committing my changes there.
> I will ask later how to get those changes into your repository via a pull request (which I have also never done before).

Almost 6 month since your first Quote. Is there any progress? Is there any assistance to coordinate the job?
What are the reachable goals to get a basic documentation to start off?

**Comment by FloVanGH** [comment]
One of the next topics are the simplification of some API parts. I think that should be done before releasing a guide with source code examples. To get a little more guidance until the guide is released I’ve added base concepts to the Readme.

But I hope we could release the first iteration of the guide after the api update.

**Comment by rzerres** [comment]
@FloVanGH 
Thanks  for the fast answer. I do appreciate all your valuable work and will try to dig deeper into the example code for now.
I will use the API to implement a basic from based application. 

For what i have seen so for, one missing peace is the absence of a 'Login' widget, beside usefull stuff like "Tree View, Table View". The latter already seems to be on the agenda.

**Comment by rzerres** [comment]
@FloVanGH 

I updated some comments as an issue for the "Do it" reference application. You will find it at.

https://codeberg.org/flovanco/doit/issues/7

**Comment by FloVanGH** [comment]
https://github.com/redox-os/orbtk/issues/108
